JAVA-3070: retryable reads

jstewart148 opened the pull request:
This is the first part of the implementation of retryable reads. The operation tests that currently pass in the sync test runner include `count`, `distinct`, `find`, `findOne`, `aggregate`, `listDatabases` and `listDatabaseNames`.

jyemin committed:
Add change stream specification test for consecutive resume

JAVA-3231

jyemin committed:
Bump to 3.11.0-beta2

jyemin committed:
Bump to 3.11.0-SNAPSHOT

jyemin committed:
Un-deprecate SocketStreamFactory constructor that takes a SocketFactory

There may be some use cases for this that we have not considered, e.g.
an ssh proxy.  So undeprecating until we know more.

JAVA-3252

jyemin committed:
Un-deprecate org.bson.Transformer

It's still used in several Codec implementations, so it should not have
been deprecated.

JAVA-3234

jyemin committed:
Deprecate meta-operator constants in com.mongodb.QueryOperators

JAVA-3224

jyemin committed:
Annotate minDistance/maxDistance parameters as Nullable in Filters class

JAVA-3223

rozza committed:
Ensure srv resolution works for legacy

JAVA-3251

rozza committed:
Ensure srv dns test runs on java 7+

JAVA-3251

jyemin committed:
Get server connectionId from isMaster if possible

As of MongoDB 4.2, the isMaster response includes the connectionId.
Formerly this value was only available in the response to the
getLastError command.  If the driver finds the connectionId in the
isMaster response, it will use it and then avoid executing the
getLastError command.

JAVA-3255

rozza committed:
Replaced legacy InitialDnsSeedlistTest with a connectivity test

JAVA-3251

jstewart148 committed:
JAVA-3070: retryable reads

jyemin committed:
Treat unrecognized boolean value in connection string as unset

Change behavior of an unrecognized boolean value in the connection
string to be as if the option was unset.

For legacy compatibility, "true", "yes", and "1" still indicate true,
and "false", "no", and "0" indicate false.

JAVA-3232

rozza committed:
Ensure use custom client for connectivity test

JAVA-3251

jyemin commented on driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java:
```diff
@@ -71,7 +71,8 @@
         super(clientSession, executor, readConcern, readPreference);
-        this.operations = new AsyncOperations<TDocument>(namespace, documentClass, readPreference, codecRegistry, writeConcern, false);
+        this.operations = new AsyncOperations<TDocument>(namespace, documentClass, readPreference, codecRegistry, writeConcern,
+                true, true);
```
Why change retryWrites to true?

jyemin replied:
It might actually be correct, in that with a $out stage an aggregation is actually a write operation.  But if so, the value needs to be plumbed in through the constructor.  Let's handle that in a separate commit.

jstewart148 replied:
Ok. I have reverted back to `false`.

jyemin commented on driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java:
```diff
@@ -71,7 +71,8 @@
         super(clientSession, executor, readConcern, readPreference);
-        this.operations = new AsyncOperations<TDocument>(namespace, documentClass, readPreference, codecRegistry, writeConcern, false);
+        this.operations = new AsyncOperations<TDocument>(namespace, documentClass, readPreference, codecRegistry, writeConcern,
+                true, true);
```
Will retryReads be plumbed through in the constructor?

jstewart148 replied:
I have opted to use a fluent setter for the Iterables.

jyemin replied:
I'm starting to see that we're inconsistent about when we use constructor params and when we use fluent setters.  In the past, we were more likely to use the fluent setters for public classes (like the Read/WriteOperation implementations.  But that distinction is going away since those classes are no longer going to be public in the 4.x driver.

For now, let's be consistent at least per-package.  So if everything is currently a constructor parameter, make the new properties constructor parameters.  And if everything is a fluent setter, make the new properties fluent setters as well.

Sorry for the confusion around this.

jstewart148 replied:
I have followed the guideline. However, there are now some constructors that contain more than 11 parameters with the addition of the `retryReads` parameter, which generates errors when running `./gradlew driver-sync:checkstyleMain`.

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/Fixture.java:
```diff
@@ -65,7 +69,14 @@ public static MongoClientSettings getMongoClientSettings() {
     public static com.mongodb.MongoClientSettings.Builder getMongoClientBuilderFromConnectionString() {
         com.mongodb.MongoClientSettings.Builder builder = com.mongodb.MongoClientSettings.builder()
-                .applyConnectionString(getConnectionString());
+                .applyConnectionString(getConnectionString())
```
I don't think we need to make this change in the global MongoClient.  Does it help?

jstewart148 replied:
I reverted this back to its original state.

jyemin commented on driver-core/src/main/com/mongodb/ConnectionString.java:
```diff
@@ -1157,6 +1164,17 @@ public Boolean getRetryWritesValue() {
+     * @since 3.11
+     * @mongodb.server.release 3.6
+     */
+    public Boolean getRetryReads() {
```
Let the MongoClient determine the default if it's unset, and just return `retryReads` value directly here.

jstewart148 replied:
Done.

jyemin commented on driver-core/src/main/com/mongodb/operation/AggregateExplainOperation.java:
```diff
@@ -51,14 +51,20 @@
+        this(namespace, pipeline, true);
+    }
+
+    AggregateExplainOperation(final MongoNamespace namespace, final List<BsonDocument> pipeline, final boolean retryReads) {
```
For all the operations, make retryReads a property with its own fluent setter, rather than overload all the constructors.  We generally reserve constructor parameters for properties that are required for execution, and not ones with sensible defaults. That avoids constructor churn in all the operations as we add new properties over time

jstewart148 replied:
Done. I have also done this for the associated `Iterable` classes, as well as for `Operations`, `SyncOperations` and `AsyncOperations`.

jyemin replied:
See my comment above about when to use fluent setters and when to use constructor parameters

jstewart148 replied:
Done.

jyemin commented on driver-core/src/main/com/mongodb/operation/AggregateExplainOperation.java:
```diff
@@ -153,7 +159,7 @@ public BsonDocument execute(final ReadBinding binding) {
             public BsonDocument call(final Connection connection) {
                 validateCollation(connection, collation);
-                return executeWrappedCommandProtocol(binding, namespace.getDatabaseName(), getCommand(), connection);
+                return CommandOperationHelper.executeCommand(binding, retryReads, namespace.getDatabaseName(), getCommand(), connection);
```
Nit: statically import the method

jstewart148 replied:
Done. IntelliJ did this when refactoring, so it occurred in a number of files.

jyemin commented on driver-core/src/main/com/mongodb/operation/AggregateExplainOperation.java:
```diff
@@ -153,7 +159,7 @@ public BsonDocument execute(final ReadBinding binding) {
             public BsonDocument call(final Connection connection) {
                 validateCollation(connection, collation);
-                return executeWrappedCommandProtocol(binding, namespace.getDatabaseName(), getCommand(), connection);
+                return CommandOperationHelper.executeCommand(binding, retryReads, namespace.getDatabaseName(), getCommand(), connection);
```
Nit: make retryReads the last parameter since it's kind of the optional one.  (You can use IntelliJ's Change Signature auto-refactoring to make this change)

jstewart148 replied:
Done.

jyemin commented on driver-core/src/main/com/mongodb/operation/CommandOperationHelper.java:
```diff
@@ -105,47 +107,108 @@ public BsonDocument apply(final BsonDocument command) {
 
-    static <D, T> T executeWrappedCommandProtocol(final ReadBinding binding, final String database, final BsonDocument command,
-                                                  final Decoder<D> decoder, final CommandTransformer<D, T> transformer) {
+    static <D, T> T executeCommand(final ReadBinding binding, final boolean retryReads, final String database, final BsonDocument command,
```
Can this be DRY'd up with the overloaded method on line 176 that has the same parameters plus a Connection?

jstewart148 replied:
I have DRY'd up the method.

jyemin commented on driver-core/src/main/com/mongodb/operation/CommandOperationHelper.java:
```diff
@@ -105,47 +107,108 @@ public BsonDocument apply(final BsonDocument command) {
+        MongoException exception;
+
+        try {
+            return executeWrappedCommandProtocol(database, command, decoder, connection, binding.getReadPreference(), transformer,
```
Nit: rename all the executeWrappedCommandProtocol methods to just executeCommand

jstewart148 replied:
Done.

jyemin commented on driver-core/src/main/com/mongodb/operation/CountOperation.java:
```diff
@@ -235,8 +258,8 @@ public Long execute(final ReadBinding binding) {
                 @Override
                 public Long call(final Connection connection) {
                     validateReadConcernAndCollation(connection, binding.getSessionContext().getReadConcern(), collation);
```
This is a bit of a problem.  The reason CountOperation (and others) first get a Connection before calling executeCommand is so that it can validate properties against the server version reported by the Connection (in this case readConcern and collation).  But on the retry attempt we get a new Connection, perhaps to a server running a different version of mongodb, and these properties have to be re-validated against that connection's reported version.  

Other operations don't just validate, but also change the command that they create (like adding a property only if the server is a certain version.

So I think we're going to need another layer of indirection here: instead of passing the actual BsonDocument of the command to `executeCommand`, we're going to have to pass a `CommandCreator` in, similar to what we do in `executeRetryableCommand` methods for retryable writes.  That way, the command creator can be called for each retry attempt, and the properties can be revalidated and the command re-created based on server version.

It also occurs to me that with this approach, CountOperation and perhaps others don't need to get a Connection first at all.  Instead they can just call the `executeCommand` method that just take a ReadBinding and no Connection, and when `executeCommand` calls the `CommandCreator` it passes the `ServerDescription` and `ConnectionDescription`, and at that time the `CommandCreator` for the operation can do it's validation, etc. using those.

jstewart148 replied:
For the operations that I have changed, each has a new `getCommandCreator` method that returns the `getCommand` method in the `CommandCreator#create` method.

jyemin reviewed (changes requested):
It's looking really good.  I like how the operations barely need to change to do this.

I did find one subtle bug that will require a change in the design.  See my comment on CountOperation, which is where I first spotted it.

jstewart148 committed:
JAVA-3070: code review changes

jstewart148 committed:
JAVA-3249: countDocuments should use group with _id: 1

jstewart148 committed:
JAVA-3070: remove fluent setters from Iterable classes
JAVA-3249: countDocuments should use group with _id: 1

jstewart-mongo commented:
With the latest commit, all of the retryable-reads tests pass. Additionally, `./gradlew driver-sync:test` runs without error. Most of the tests pass when running `./gradlew driver-core:test`. The ones that fail are async tests, and I have not started on the async modifications.
I have also added the change for JAVA-3249.

jstewart148 committed:
JAVA-3070: retryable reads - async and legacy

jstewart-mongo commented:
The async and legacy tests pass locally. `./gradlew check` passes as well. The latest Evergreen patch shows a number of issues left to resolve.

Latest Evergreen patch: https://evergreen.mongodb.com/version/5cae3db257e85a246058b7ca

jstewart148 committed:
JAVA-3070: update test runners to properly process runOn array

jstewart-mongo commented:
Updated the test runners to process the `runOn` array properly.
Evergreen patch: https://evergreen.mongodb.com/version/5caf83c257e85a24605cab66

jyemin committed:
Deprecated isModifiedCountAvailable methods

Additionally, remove usage of this method and simplify code to assume
that the modifiedCount is available on all supported server versions.

JAVA-3262

jstewart148 committed:
JAVA-3070: use connectionString for authentication tests in the test runner

jstewart148 committed:
JAVA-3070: static checks changes

jyemin committed:
Disable tests failing on MongoDB 4.2

* Some tests are failing due to changes in the structure of the response
  to an explain command
* One is failing due to removal of flags from response to
  listCollections command

Commit:
Remove unused imports

jstewart148 committed:
JAVA-3070: update CommandTransformer and CommandCreator usage

jstewart-mongo commented:
I have updated the CommandCreator and CommandTransformer so that 
- validation of collation and read concern can be performed when the command is created
- the correct connection is used on retry
- a fresh ConnectionSource is used when needed (for example, when allocating a new AsyncQueryBatchCursor in the `asyncTransformer` method in AggregateOperationImpl.
The latest Evergreen patch (https://evergreen.mongodb.com/version/5cb797e91e2d172251675d32) is showing a few more errors than what I am seeing locally. Continuing to investigate.

jstewart148 committed:
JAVA-3070: add CommandWriteTransformer for async write operations

jstewart-mongo commented:
Fixed write transformer issues in async by adding a `CommandWriteTransformer` interface for write operations. Evergreen tests for latest shows the expected failures.
https://evergreen.mongodb.com/version/5cb7c079a4cf4744bfa562f1

jyemin committed:
Fix checkstyle warning

jyemin committed:
Fix find command tests for min/max by adding a hint.

MongoDB 4.2 will require that any use of min or max for the find command
hints use of an index on the field

jstewart148 committed:
JAVA-3070: fix driver-core tests

jstewart148 committed:
JAVA-3070: refactoring of CommandTransformer

jstewart148 committed:
JAVA-3070: sync changes to transformer

jstewart148 committed:
JAVA-3070: fix ParallelCollectionScanOperation test

jstewart-mongo commented:
I have made the changes to the operations using a bad ConnectionSource on retry. I have also simplified some of the operations' `execute` methods. All tests pass locally, and the latest Evergreen patch is all green (https://evergreen.mongodb.com/version/5cb9de09d1fe0711c96235cb). Moving this to formal code review.

jstewart-mongo commented:
After a recent rebase, the PR became too cluttered with other commits to properly review. I opened a new PR to continue this review: https://github.com/jstewart-mongo/mongo-java-driver/pull/26

rozza commented:
@jstewart-mongo I think this PR can be closed!

jyemin commented:
Closing!
